File: src/main/scala/arimitsu/sf/cql/v3/messages/Startup.scala
package arimitsu.sf.cql.v3.messages

import arimitsu.sf.cql.v3._
import arimitsu.sf.cql.v3.util.{BufferUtil, Notation}
import java.nio.ByteBuffer
import scala.Some

/**
 * Created by sxend on 14/06/07.
 */
case class Startup(streamId: Short, flags: Byte, compression: Option[Compression]) extends Request {
  val default = Map(
    "CQL_VERSION" -> "3.0.0"
  )

  override def toFrame: Frame = {
    val map = compression match {
      case Some(s) => default + ("COMPRESSION" -> s.name)
      case None => default
    }
    val notationResult = Notation.toStringMap(map)
    val buffer = ByteBuffer.wrap(notationResult)
    val length = BufferUtil.length(buffer)
    Frame(Header(Version.REQUEST, flags, streamId, Opcode.STARTUP),
      length,
      Option(buffer))
  }
}

File: src/main/scala/arimitsu/sf/cql/v3/messages/Startup.java
package arimitsu.sf.cql.v3.messages;

import arimitsu.sf.cql.v3.*;
import arimitsu.sf.cql.v3.util.Notation;

import java.util.Map;

/**
 * Created by sxend on 14/06/07.
 */
public class Startup implements Request {
    public static final String CQL_VERSION = "CQL_VERSION";
    public static final String CQL_VERSION_NUMBER = "3.0.0";
    public static final String OPTION_COMPRESSION = "COMPRESSION";
    public final short streamId;
    public final Flags flags;
    public final Map<String, String> options;

    public Startup(short streamId, Flags flags, Map<String, String> options) {
        this.streamId = streamId;
        this.flags = flags;
        this.options = options;
    }

    @Override
    public Frame toFrame() {
        byte[] body = Notation.toStringMap(options);
        return new Frame(new Header(Version.REQUEST, flags, streamId, Opcode.STARTUP), body);
    }

}
